javalib/src/main/scala/java/util/Base64.scala

```diff
@@ -0,0 +1,486 @@
+    }
+    new Encoder(basicEncodeTable, lineLength / 4 * 4, lineSeparator)
+  }
+
+  def getDecoder(): Decoder = basicDecoder
+
+  def getUrlDecoder(): Decoder = urlSafeDecoder
+
+  def getMimeDecoder(): Decoder = mimeDecoder
+
+  // --------------------------------------------------------------------------
+
+  class Decoder private[Base64](table: Array[Int], ignoreInvalid: Boolean) {
+
+    def decode(src: Array[Byte]): Array[Byte] = {
+      val dst = new Array[Byte](dstRequiredLength(src, 0, src.length))
+      doDecode(new Wrapper(src), new Wrapper(dst))
+      dst
+    }
+
+    def decode(src: String): Array[Byte] = decode(src.getBytes(StandardCharsets.ISO_8859_1))
+
+    def decode(src: Array[Byte], dst: Array[Byte]): Int = {
+      if (dst.length >= dstMaxLength(src.length)) // dst is definitely large enough
+        doDecode(new Wrapper(src), new Wrapper(dst))
+      else if (dst.length < dstRequiredLength(src, 0, src.length))
+        throw new IllegalArgumentException("Output byte array is too small for decoding all input bytes")
+      else
+        doDecode(new Wrapper(src), new Wrapper(dst))
+    }
+
+    def decode(buffer: ByteBuffer): ByteBuffer = {
+      val start = buffer.position()
+      try {
+        val src = buffer.array()
+        val offset = buffer.arrayOffset() + buffer.position()
+        val length = buffer.arrayOffset() + buffer.limit()
+        val dst = new Array[Byte](dstRequiredLength(src, offset, length))
+        val written = doDecode(new Wrapper(src, offset, length), new Wrapper(dst))
+        ByteBuffer.wrap(dst, 0, written)
+      } catch {
+        case e: IllegalArgumentException =>
+          buffer.position(start)
+          throw e
+      }
+    }
+
+    def wrap(is: InputStream): InputStream = new DecodingInputStream(is, table, ignoreInvalid)
+
+    // ------------------------------------------------------------------------
+    // PRIVATE
+    // ------------------------------------------------------------------------
+
+    private def doDecode(src: Wrapper[Byte], dst: Wrapper[Byte]): Int = {
+      val srcBuffer = new Wrapper[Int](new Array[Int](4))
+
+      @inline
+      def inputData(): Unit = {
+        srcBuffer.position = 0
+        var shift = 18
+        var i = 0
+        while(srcBuffer.hasRemaining) {
+          i |= (srcBuffer.get << shift)
+          shift -= 6
+        }
+        if(shift == 12) throw new IllegalArgumentException("Last unit does not have enough valid bits")
+        if(shift <= 6) dst.put((i >> 16).toByte)
+        if(shift <= 0) dst.put((i >> 8).toByte)
+        if(shift <= -6) dst.put(i.toByte)
+        srcBuffer.clear()
+      }
+
+      @inline
+      def iterate(): Unit = {
+        // scalastyle:off return
+        while (src.hasRemaining) {
+          if (srcBuffer.hasRemaining) {
+            val int = src.get & 0xff
+            table(int) match {
+              case -2 => return
```

Comment:
You can get rid of this `return` by turning `iterate()` into a tail-recursive method.